packages/backend/src/league/league_controller.ts

```diff
@@ -265,4 +268,62 @@ export class LeaguesController extends ControllerWrap<League> {
       }),
     );
   }
+
+  /**
+   * Generates match proposal
+   */
+  @Post("{leagueId}/generate-match-proposals")
+  public async generateMatchProposals(
+    @Path() leagueId: ID,
+    @Request() req: express.Request,
+  ): Promise<WithError<MatchProposal[]>> {
+    const id = new ObjectId(leagueId);
+
+    return this.withUser(req, async (currUser) => {
+      const res = await this.get(id);
+      if (!res.success) {
+        return res;
+      }
+
+      const league = res.data;
+      if (!hasId(league.ownerIds, currUser._id)) {
+        return this.notFound();
+      }
+
+      const matchProposals: MatchProposal[] = [];
+
+      // Group people by ability level
+      const peopleByAbilityLevel: Record<AbilityLevel, ObjectId[]> = {
+        beginner: [],
+        intermediate: [],
+        advanced: [],
+      };
+      for (const person of league.people as User[]) {
+        peopleByAbilityLevel[person.abilityLevel].push(person._id);
+      }
+
+      // Generate match proposals for each ability level
+      for (const abilityLevel of [
+        "beginner",
+        "intermediate",
+        "advanced",
+      ] as const) {
```

Comment:
Please iterate over the enum `AbilityLevel`, therefore `as const` won't be required

Reply:
do you mean: for (const abilityLevel in AbilityLevel) {
  const people = peopleByAbilityLevel[abilityLevel as keyof typeof AbilityLevel];
  // ...
}

Reply:
Yup you can apparently do:

```ts
Object.fromEntries(
  Object
    .keys(AbilityLevel)
    .map((k) => [k, []])
)
```

Note here, we are using `Object.keys` which gives you the enum basically as an array: `["beginner"...]` and converting it to a tuple of `[string, unknown[]]`, which is the form `Object.fromEntries` takes to create an object with keys that are located in index `0` (so the strings) and values in index `1` (so the `[]`).

You will also have to specify the type which as you've noted is

```ts
Record<keyof typeof AbilityLevel, ObjectId[]>
```

Reply:
Oh oops I realised I have given you the code for https://github.com/ESP-Number-One/MatchyoSports/pull/91#discussion_r1521676804 and not this comment

You code there should work perfectly fine